File: src/python/subscriber.py
import zmq
import numpy as np
import cv2
import time
import json  # Add import for JSON handling

class Subscriber:
    def __init__(self, address="tcp://127.0.0.1:5555", topic="camera/image_raw"):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.SUB)
        self.socket.connect(address)
        self.socket.setsockopt_string(zmq.SUBSCRIBE, topic)
        
        self.image_chunks = {}

    def receive_message(self, timeout=None):
        """ Recibe mensajes binarios que contienen múltiples imágenes y datos adicionales. """
        self.total_bytes_received = 0  # Initialize total bytes received
        start_time = time.time()  # Start time for reception

        if timeout:
            self.socket.setsockopt(zmq.RCVTIMEO, timeout)

        while True:
            try:
                # print("Esperando recibir un mensaje...")
                topic, index, total, chunk = self.socket.recv_multipart()
                # print(f"Mensaje recibido: index={index}, total={total}")
            except zmq.Again:
                print("❌ Timeout: No se recibieron mensajes en el tiempo esperado.")
                return None, None

            index = int(index.decode())
            total = int(total.decode())

            self.image_chunks[index] = chunk
            self.total_bytes_received += len(chunk)

            if len(self.image_chunks) == total:  # Si todos los fragmentos llegaron
                ordered_chunks = [self.image_chunks[i] for i in sorted(self.image_chunks.keys())]
                message_bytes = b"".join(ordered_chunks)

                # Split the JSON part and the image part
                if b'\x00' in message_bytes:
                    json_part, images_bytes = message_bytes.split(b'\x00', 1)
                else:
                    json_part = message_bytes
                    images_bytes = b''

                try:
                    # print("Decodificando JSON...")
                    message = json.loads(json_part.decode('utf-8'))
                    # print(f"📄 Mensaje decodificado: {message}")
                    images = []
                    if "images" in message and message["images"]:
                        offset = 0
                        for image_info in message["images"]:
                            size = image_info["metadata"]["size"]
                            format = image_info["metadata"]["format"]
                            dtype = image_info["metadata"]["dtype"]
                            width = image_info["metadata"]["width"]
                            height = image_info["metadata"]["height"]
                            channels = image_info["metadata"]["channels"]
                            image_bytes = images_bytes[offset:offset + size]
                            if format == "raw":
                                image_array = np.frombuffer(image_bytes, dtype=dtype).reshape((height, width, channels))
                            elif format in ["jpeg", "png"]:
                                image_array = cv2.imdecode(np.frombuffer(image_bytes, dtype=np.uint8), cv2.IMREAD_COLOR)
                            else:
                                print(f"❌ Formato de imagen no soportado: {format}")
                                self.image_chunks.clear()
                                return None, None
                            images.append(image_array)
                            offset += size
                        print(f"✅ Imágenes embebidas en binario reconstruidas correctamente.")
                    self.image_chunks.clear()  # Limpiar para el siguiente mensaje
                    return images, message.get("data", {})
                except json.JSONDecodeError:
                    print("❌ Error decodificando JSON.")
                    self.image_chunks.clear()
                    return None, None

    def close(self):
        """ Cierra la conexión ZeroMQ. """
        self.socket.close()
        self.context.term()

File: src/cpp/src/subscriber.cpp
#include "subscriber.hpp"
#include <iostream>
#include <nlohmann/json.hpp>
#include <unordered_map>
#include <variant>
#include <opencv2/imgcodecs.hpp>

using json = nlohmann::json;


Subscriber::Subscriber(const std::string &address, const std::string &topic)
    : context(1), socket(context, zmq::socket_type::sub) {
    socket.connect(address);
    socket.set(zmq::sockopt::subscribe, topic);
    std::cout << "✅ Subscriber C++ conectado a " << address << std::endl;
}

std::pair<std::vector<cv::Mat>, DataMap> Subscriber::receive_message(int timeout_ms) {
    std::string topic_str;
    std::vector<std::string> chunks;
    int num_chunks_int = -1;

    while (true) {
        zmq::message_t topic, index, num_chunks, chunk;

        auto res1 = socket.recv(topic, zmq::recv_flags::none);
        if (!res1) {
            throw std::runtime_error("❌ Timeout: No se recibieron mensajes en el tiempo esperado.");
        }
        
        auto res2 = socket.recv(index, zmq::recv_flags::none);
        auto res3 = socket.recv(num_chunks, zmq::recv_flags::none);
        auto res4 = socket.recv(chunk, zmq::recv_flags::none);

        if (!res2 || !res3 || !res4) {
            throw std::runtime_error("❌ Error recibiendo partes del mensaje multipart");
        }

        // Extraer los valores como strings
        std::string index_str(static_cast<char*>(index.data()), index.size());
        std::string num_chunks_str(static_cast<char*>(num_chunks.data()), num_chunks.size());

        int index_int = std::stoi(index_str);
        num_chunks_int = std::stoi(num_chunks_str);

        if (topic_str.empty()) {
            topic_str.assign(static_cast<char*>(topic.data()), topic.size());
            chunks.resize(num_chunks_int);
        }

        // Verificar que el mensaje es consistente (mismo topic y tamaño esperado)
        if (topic_str != std::string(static_cast<char*>(topic.data()), topic.size())) {
            throw std::runtime_error("❌ Error: Cambio de topic en mitad del mensaje.");
        }

        if (index_int < 0 || index_int >= num_chunks_int) {
            throw std::runtime_error("❌ Error: Índice fuera de rango.");
        }

        // Guardar el chunk en su posición
        chunks[index_int].assign(static_cast<char*>(chunk.data()), chunk.size());

        // Si ya tenemos todos los chunks, salir del bucle
        bool mensaje_completo = true;
        for (int i = 0; i < num_chunks_int; ++i) {
            if (chunks[i].empty()) {
                mensaje_completo = false;
                break;
            }
        }

        if (mensaje_completo) {
            break;
        }
    }

    // Concatenar todos los chunks en uno solo
    std::string messageBytes;
    for (const auto& chunk : chunks) {
        messageBytes += chunk;
    }

    // Ahora sigue igual que antes...
    size_t pos = messageBytes.find('\0');
    if (pos == std::string::npos) {
        throw std::runtime_error("❌ Error: separador '\\0' no encontrado en el mensaje");
    }

    std::string header_json = messageBytes.substr(0, pos);
    std::string images_data = messageBytes.substr(pos + 1);

    json message_json = json::parse(header_json);

    DataMap data;
    for (auto& [k, v] : message_json["data"].items()) {
        if (v.is_string()) {
            data[k] = v.get<std::string>();
        } else if (v.is_number_integer()) {
            data[k] = v.get<int>();
        } else if (v.is_number_float()) {
            data[k] = v.get<double>();
        } else if (v.is_boolean()) {
            data[k] = v.get<bool>();
        } else if (v.is_null()) {
            data[k] = nullptr;
        } else {
            data[k] = v.dump();
        }
    }

    std::vector<cv::Mat> images;
    int offset = 0;

    for (const auto& img_meta : message_json["images"]) {
        int width = img_meta["metadata"]["width"];
        int height = img_meta["metadata"]["height"];
        int channels = img_meta["metadata"]["channels"];
        int size = img_meta["metadata"]["size"];
        std::string dtype = img_meta["metadata"]["dtype"];
        std::string format = img_meta["metadata"]["format"];

        int type = numpy_dtype_to_opencv(dtype, channels);

        if (offset + size > images_data.size()) {
            std::cerr << "❌ Error: Esperando " << size 
            << " bytes, pero solo hay " << (images_data.size() - offset) 
            << " disponibles." << std::endl;
            throw std::runtime_error("❌ Error: Datos de imagen incompletos.");
        }

        cv::Mat img;
        if (format == "raw") {
            img = cv::Mat(height, width, type);
            std::memcpy(img.data, images_data.data() + offset, size);
        } else if (format == "jpeg" || format == "png") {
            std::vector<uchar> buffer(images_data.begin() + offset, images_data.begin() + offset + size);
            img = cv::imdecode(buffer, cv::IMREAD_COLOR);
        } else {
            throw std::runtime_error("❌ Error: Formato de imagen no soportado: " + format);
        }

        images.push_back(img);
    }

    return {images, data};
}

int Subscriber::numpy_dtype_to_opencv(const std::string& dtype, int channels) {
    int type;
    if (dtype == "uint8") type = CV_8UC(channels);
    else if (dtype == "int8") type = CV_8SC(channels);
    else if (dtype == "uint16") type = CV_16UC(channels);
    else if (dtype == "int16") type = CV_16SC(channels);
    else if (dtype == "float32") type = CV_32FC(channels);
    else if (dtype == "float64") type = CV_64FC(channels);
    else throw std::runtime_error("Tipo de imagen no soportado: " + dtype);
    return type;
}
